keps/sig-cluster-lifecycle/kubeadm/20191125-remove-clusterstatus-from-kubeadm-config.md

```diff
@@ -0,0 +1,138 @@
+
+This approach does not require the maintenance of a separated list, and implicitly always reflect the current status of the cluster.
+
+This allows to remove the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and to clean-up all the related code in `kubeadm init`, `kubeadm join` and `kubeadm reset`
+
+### Goals
+
+- To introduce a new method for tracking the list of API endpoints in a cluster
+- To allow removal of the `ClusterStatus` entry in the `kubeadm-config` ConfigMap and clean-up of the related goal.
+
+### Non-Goals
+
+- To change any user facing behavior in `kubeadm`
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+As of today, the `ClusterStatus` entry in the `kubeadm-config` ConfigMap contains a map that stores the `LocalAPIEndpoint` for each control-plane node.
+
+The `LocalAPIEndpoint` primary usage is for the `advertise-address` flag in the `kube-apiserver` pod. Having this value in a flag is not ideal for the purpose of this proposal, so, we are going to echo the same value into a new annotation named `kubeadm.kubernetes.io/kube-apiserver.advertise-address`.
+
+Once the annotation will be in place, it will be possible to easily retrieve the local advertise address for each control plane node by querying the corresponding `kube-apiserver` pod.
+
+The `LocalAPIEndpoint` is also used in the stacked `etcd` pod manifest for composing the `peer-urls` and the `client-urls`; the latter is used by kubeadm when accessing etcd in an existing cluster, e.g. when doing `join --control-plane`.
+
+We are going to echo the `client-urls` value into a new annotation named `kubeadm.kubernetes.io/etcd.advertise-client-urls`. Once the annotation will be in place, it will be possible to easily retrieve the etcd client urls by querying the `etcd` pods.
+
+### Risks and Mitigations
+
+R. The list of API endpoints in a cluster is crucial to all the kubeadm workflows
+M. The proper function of those workflows and of the underlying codes is already covered by E2E tests; on top of that, we are going to try to implement this change at the beginning of the v1.18 cycle, thus ensuring as much
+test cycles as possible.
+
+## Design Details
+
+### Test Plan
+
+No additional test E2E test are required for this change because all the affected behaviors are already covered by existing E2E test.
+
+Additional unit test are required only for the new function implementing the inspection of the current Pods.
+
+### Graduation Criteria
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+During upgrades:
+
+- The new annotations `kubeadm.kubernetes.io/kube-apiserver.advertise-address` and `kubeadm.kubernetes.io/etcd.advertise-client-urls` will be generated during the upgrade of the static pod manifests.
+- The `ClusterStatus` entry will be cleaned up during the upgrade of the `kubeadm-config` ConfigMap.
```

Comment:
We need to guarantee that `ClusterStatus`-es are kept together with annotations for some time (they can be considered beta, so at least 9 months/3 cycles) not to break people.

Reply:
The whole point of this KEP is that it is not possible to keep ClusterStatus up to date with the current status.
see [1]

Reply:
we discussed during kubeadm office hours that we would like to keep the ClusterStatus for a while.
@fabriziopandini please confirm. does this section needs a minor update?